### PR TITLE
:sparkles: refactor and update Makefile

### LINKED ISSUE #5574: Makefile targets re-shuffle
The CAPI Makefile is getting bigger, and we should consider if/how to make things easier to use for the new comers and for the existing maintainers as well.

A possible set of actions to make this happen using the current machinery in place is to:

- [ ] standardise targets names (e.g. some of them use kcp/cabpk/capd suffix, some other the full name of the same components)
- [ ] standardise targets documentation (e.g all with or without a final dot, consistent wording etc. etc)
- [ ] ensure all the make targets are documented/available (e.g. not all the verify checks have a dedicated target, same for hack tools)
- [ ] re shuffle make targets order, so that we can group things in sections by adding comments starting with ##@ 

Note: activities can go in separated PRs if this helps in making this happen.

A mock of a possible outcome of this activity is:

Screenshot 2021-11-02 at 22 41 56

NOTE: once this work is completed, then it will be easier to:
- [ ] get rid of nested Makefile in the Cluster API code base, thus simplifying dev-X and maintenance
- [ ] identify and create additional targets, e.g around starting e2e with a single command, or creating a local repository for clusterctl development

/kind feature

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Signed-off-by: Prajyot-Parab <prajyot.parab2@ibm.com>

**What this PR does / why we need it**:

1. Standardised target names for all the components.
2. Consistent documentation across all the targets.
3. Added dedicated targets for missing verify and hack/tools components.
4. Re-shuffled order, grouped targets for easy reading, understanding and maintenance.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5574

### CODE REVIEW

**Comment by stmcginnis** [comment]
/lgtm

Still need CLA signed.

Help output looks a lot better organized now:

```sh
Usage:
  make <target>

Targets:

generate:
  generate                                       Run all generate-manifests-*, generate-go-deepcopy-* and generate-go-conversions-* targets
  generate-manifests                             Run all generate-manifests-* targets
  generate-manifests-core                        Generate manifests e.g. CRD, RBAC etc. for core
  generate-manifests-kubeadm-bootstrap           Generate manifests e.g. CRD, RBAC etc. for kubeadm bootstrap
  generate-manifests-kubeadm-control-plane       Generate manifests e.g. CRD, RBAC etc. for kubeadm control plane
  generate-go-deepcopy                           Run all generate-go-deepcopy-* targets
  generate-go-deepcopy-core                      Generate deepcopy go code for core
  generate-go-deepcopy-kubeadm-bootstrap         Generate deepcopy go code for kubeadm bootstrap
  generate-go-deepcopy-kubeadm-control-plane     Generate deepcopy go code for kubeadm control plane
  generate-go-conversions                        Run all generate-go-conversions-* targets
  generate-go-conversions-core                   Generate conversions go code for core
  generate-go-conversions-kubeadm-bootstrap      Generate conversions go code for kubeadm bootstrap
  generate-go-conversions-kubeadm-control-plane  Generate conversions go code for kubeadm control plane
  diagrams                                       Generate diagrams for *.plantuml files

lint and verify:
  modules                                        Run go mod tidy to ensure modules are up to date
  lint                                           Lint the codebase
  lint-fix                                       Lint the codebase and run auto-fixers if supported by the linter
  apidiff                                        Check for API differences
  format-tiltfile                                Format the Tiltfile
  verify                                         Run all verify-* targets
  verify-modules                                 Verify go modules are up to date
  verify-gen                                     Verfiy go generated files are up to date
  verify-conversions                             Verifies API expected conversion are in place
  verify-boilerplate                             Verify boilerplate text exists in each file
  verify-shellcheck                              Verify shell files
  verify-tiltfile                                Verify Tiltfile format

build:
  clusterctl                                     Build the clusterctl binary
  managers                                       Run all manager-* targets
  manager-core                                   Build the core manager binary into the ./bin folder
  manager-kubeadm-bootstrap                      Build the kubeadm bootstrap manager binary into the ./bin folder
  manager-kubeadm-control-plane                  Build the kubeadm control plane manager binary into the ./bin folder
  docker-build                                   Run docker-build-core and all docker-build-kubeadm-* targets
  docker-build-all                               Build all the architecture docker images
  docker-build-core                              Build the docker image for core controller manager
  docker-build-kubeadm-bootstrap                 Build the docker image for kubeadm bootstrap controller manager
  docker-build-kubeadm-control-plane             Build the docker image for kubeadm control plane controller manager
  e2e-framework                                  Builds the CAPI e2e framework

test:
  test                                           Run all the tests
  test-e2e                                       Run the e2e tests
  test-junit                                     Run tests with verbose setting and generate a junit report
  test-cover                                     Run tests with code coverage and code generate reports
  test-verbose                                   Run tests with verbose settings
  serve-book                                     Build and serve the book with live-reloading enabled
  kind-cluster                                   Create a new kind cluster designed for testing with Tilt
  docker-build-e2e                               Rebuild all Cluster API provider images to be used in the e2e tests

release:
  release                                        Build and push container images using the latest git tag for the commit
  release-manifests                              Build the manifests to publish with a release
  release-manifests-dev                          Build the development manifests and copies them in the release folder
  release-binaries                               Build the binaries to publish with a release
  release-staging                                Build and push container images to the staging bucket
  release-staging-nightly                        Tag and push container images to the staging bucket. Example image tag: cluster-api-controller:nightly_main_20210121
  release-alias-tag                              Add the tag to the last build tag
  docker-push                                    Push the docker images
  docker-push-all                                Push all the architecture docker images
  docker-push-manifest-core                      Push the multiarch manifest for the core docker images
  docker-push-manifest-kubeadm-bootstrap         Push the multiarch manifest for the the kubeadm bootstrap docker images
  docker-push-manifest-kubeadm-control-plane     Push the multiarch manifest for the the kubeadm control plane docker images

clean:
  clean                                          Remove all generated files
  clean-bin                                      Remove all generated binaries
  clean-book                                     Remove all generated GitBook files
  clean-release                                  Remove the release folder
  clean-release-git                              Restores the git files usually modified during a release
  clean-generated-conversions                    Remove files generated by conversion-gen from the mentioned dirs. Example SRC_DIRS="./api/v1alpha4"

hack/tools:
  controller-gen                                 Build a local copy of controller-gen
  conversion-gen                                 Build a local copy of conversion-gen
  conversion-verifier                            Build a local copy of conversion-verifier
  gotestsum                                      Build a local copy of gotestsum
  go-apidiff                                     Build a local copy of apidiff
  golangci-lint                                  Build a local copy of golangci-lint
  envsubst                                       Build a local copy of envsubst
  kustomize                                      Build a local copy of kustomize
  setup-envtest                                  Build a local copy of setup-envtest
  yq                                             Build a local copy of yq
  kpromo                                         Build a local copy of kpromo
```

Current output is just one ungrouped list of all targets.

**Comment by Prajyot-Parab** [comment]
@stmcginnis Thanks for the review, I already have `cla/linuxfoundation` signed and label for same is already added. For EasyCLA I was told its a non-blocking and currently in testing phase, so they are reluctant to add my ID there until it is in stable phase. (I can still try asking them, if that will be a blocker in merging this PR)

File: Makefile [hunk 1]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+apidiff: $(GO_APIDIFF) ## Check for API differences
+	$(GO_APIDIFF) $(shell git rev-parse origin/main) --print-compatible
+
+.PHONY: format-tiltfile
+format-tiltfile: ## Format the Tiltfile
+	./hack/verify-starlark.sh fix
+
+.PHONY: verify
+verify: ## Run all verify-* targets
+	./hack/verify-doctoc.sh
```
Comment (fabriziopandini) [inline Makefile:317]: Should this have its own make target?

File: Makefile [hunk 2]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+
+apidiff: $(GO_APIDIFF) ## Check for API differences
+	$(GO_APIDIFF) $(shell git rev-parse origin/main) --print-compatible
+
+.PHONY: format-tiltfile
+format-tiltfile: ## Format the Tiltfile
+	./hack/verify-starlark.sh fix
+
+.PHONY: verify
+verify: ## Run all verify-* targets
```
Comment (fabriziopandini) [inline Makefile:316]: What about adding `ALL_VERIFY_CHECKS` var and use `$(addprefix verify-,$(ALL_VERIFY_CHECKS))`

File: Makefile [hunk 3]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+
+.PHONY: clusterctl
+clusterctl: ## Build the clusterctl binary
+	go build -trimpath -ldflags "$(LDFLAGS)" -o $(BIN_DIR)/clusterctl sigs.k8s.io/cluster-api/cmd/clusterctl
+
+.PHONY: managers
+managers: ## Run all manager-* targets
+	$(MAKE) manager-core
+	$(MAKE) manager-kubeadm-bootstrap
+	$(MAKE) manager-kubeadm-control-plane
```
Comment (fabriziopandini) [inline Makefile:383]: What about adding `ALL_MANAGERS` variable and use `$(addprefix manager-,$(ALL_MANAGERS))`

File: Makefile [hunk 4]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+manager-kubeadm-bootstrap: ## Build the kubeadm bootstrap manager binary into the ./bin folder
+	go build -trimpath -ldflags "$(LDFLAGS)" -o $(BIN_DIR)/kubeadm-bootstrap-manager sigs.k8s.io/cluster-api/bootstrap/kubeadm
+
+.PHONY: manager-kubeadm-control-plane
+manager-kubeadm-control-plane: ## Build the kubeadm control plane manager binary into the ./bin folder
+	go build -trimpath -ldflags "$(LDFLAGS)" -o $(BIN_DIR)/kubeadm-control-plane-manager sigs.k8s.io/cluster-api/controlplane/kubeadm
 
 .PHONY: docker-build
-docker-build: docker-pull-prerequisites ## Build the docker images for controller managers
+docker-build: docker-pull-prerequisites ## Run docker-build-core and all docker-build-kubeadm-* targets
```
Comment (fabriziopandini) [inline Makefile:398]: What about adding ALL_DOCKER_BUILD variable and use $(addprefix docker-build-,$(ALL_DOCKER_BUILD))

File: Makefile [hunk 5]
```diff
@@ -455,72 +424,98 @@ docker-build-kubeadm-control-plane: ## Build the docker image for kubeadm contro
...
+$(KPROMO): $(TOOLS_DIR)/go.mod # Build kpromo from tools folder.
+	cd $(TOOLS_DIR) && go build -tags=tools -o $(KPROMO_BIN) sigs.k8s.io/promo-tools/v3/cmd/kpromo
+
+$(KUSTOMIZE): # Download kustomize using hack script into tools folder.
+	hack/ensure-kustomize.sh
+
+$(GOLANGCI_LINT): .github/workflows/golangci-lint.yml # Download golanci-lint using hack script into tools folder.
+	hack/ensure-golangci-lint.sh \
+		-b $(TOOLS_DIR)/$(BIN_DIR) \
+		$(shell cat .github/workflows/golangci-lint.yml | grep version | sed 's/.*version: //')
```
Comment (fabriziopandini) [inline Makefile:464]: Should this block be moved down into the hack/tools section?

File: Makefile [hunk 6]
```diff
@@ -629,7 +624,7 @@ release-staging-nightly: ## Tags and push container images to the staging bucket
 	gsutil cp $(RELEASE_DIR)/* gs://$(STAGING_BUCKET)/components/$(NEW_RELEASE_ALIAS_TAG)
 
 .PHONY: release-alias-tag
-release-alias-tag: ## Adds the tag to the last build tag.
+release-alias-tag: ## Add the tag to the last build tag
```
Comment (fabriziopandini) [inline Makefile:627]:
```suggestion
release-alias-tag: ## Add the release alias tag to the last build tag
```

File: Makefile [hunk 7]
```diff
@@ -642,10 +637,74 @@ release-notes: $(RELEASE_NOTES_DIR) $(RELEASE_NOTES)
...
 
+## --------------------------------------
+## Docker
+## --------------------------------------
+
+.PHONY: docker-pull-prerequisites
+docker-pull-prerequisites:
+	docker pull docker.io/docker/dockerfile:1.1-experimental
+	docker pull $(GO_CONTAINER_IMAGE)
+	docker pull gcr.io/distroless/static:latest
```
Comment (fabriziopandini) [inline Makefile:648]: Should this be moved up into build given that it is used by docker-build?

**Review by fabriziopandini — COMMENTED** [review]
@Prajyot-Parab many thanks for picking up this work, I really appreciated it!
Just a couple of nits from my side, but overall changes are ok

File: Makefile [hunk 8]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+
+.PHONY: verify-gen
+verify-gen: generate  ## Verfiy go generated files are up to date
+	@if !(git diff --quiet HEAD); then \
+		git diff; \
+		echo "generated files are out of date, run make generate"; exit 1; \
+	fi
+
+.PHONY: verify-conversions
+verify-conversions: $(CONVERSION_VERIFIER)  ## Verifies API expected conversion are in place
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply Makefile]: Noted.

File: Makefile [hunk 9]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+lint: $(GOLANGCI_LINT) ## Lint the codebase
+	$(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
+	cd $(TEST_DIR); $(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
+	cd $(TOOLS_DIR); $(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
+
+.PHONY: lint-fix
+lint-fix: $(GOLANGCI_LINT) ## Lint the codebase and run auto-fixers if supported by the linter
+	GOLANGCI_LINT_EXTRA_ARGS=--fix $(MAKE) lint
+
+apidiff: $(GO_APIDIFF) ## Check for API differences
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply Makefile:309]: Noted.

File: Makefile [hunk 10]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+docker-build: docker-pull-prerequisites ## Run docker-build-core and all docker-build-kubeadm-* targets
 	$(MAKE) ARCH=$(ARCH) docker-build-core
 	$(MAKE) ARCH=$(ARCH) docker-build-kubeadm-bootstrap
 	$(MAKE) ARCH=$(ARCH) docker-build-kubeadm-control-plane
 
+.PHONY: docker-build-all
+docker-build-all: $(addprefix docker-build-,$(ALL_ARCH)) ## Build all the architecture docker images
+
+docker-build-%:
+	$(MAKE) ARCH=$* docker-build
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply Makefile]: Done.

File: Makefile (hunk 1 above)
Comment (Prajyot-Parab) [reply Makefile]: Done.

File: Makefile (hunk 2 above)
Comment (Prajyot-Parab) [reply Makefile]: +1 Added.

File: Makefile (hunk 3 above)
Comment (Prajyot-Parab) [reply Makefile]: +1 Added.

File: Makefile (hunk 4 above)
Comment (Prajyot-Parab) [reply Makefile]: +1 Added

File: Makefile (hunk 5 above)
Comment (Prajyot-Parab) [reply Makefile]: Moved, should be easier now for future contributors as related code is at same place.

File: Makefile (hunk 6 above)
Comment (Prajyot-Parab) [reply Makefile]: Noted.

File: Makefile (hunk 7 above)
Comment (Prajyot-Parab) [reply Makefile]: Makes sense, done.

File: Makefile [hunk 11]
```diff
@@ -361,82 +279,133 @@ generate-go-conversions-kcp: $(CONVERSION_GEN)
...
+	@echo "Verifying CAPD"
+	cd $(CAPD_DIR); $(MAKE) verify
 
-.PHONY: modules
-modules: ## Runs go mod to ensure modules are up to date.
-	go mod tidy
-	cd $(TOOLS_DIR); go mod tidy
-	cd $(TEST_DIR); go mod tidy
+.PHONY: verify-book-links
+verify-book-links:
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply Makefile]: Noted.

File: Makefile [hunk 12]
```diff
@@ -130,172 +130,103 @@ LDFLAGS := $(shell hack/version.sh)
...
-lint: $(GOLANGCI_LINT) ## Lint codebase
-	$(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
-	cd $(TEST_DIR); $(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
-	cd $(TOOLS_DIR); $(GOLANGCI_LINT) run -v $(GOLANGCI_LINT_EXTRA_ARGS)
+##@ generate:
 
-.PHONY: lint-fix
-lint-fix: $(GOLANGCI_LINT) ## Lint the codebase and run auto-fixers if supported by the linter.
-	GOLANGCI_LINT_EXTRA_ARGS=--fix $(MAKE) lint
+ALL_GENERATE_MODULES = core kubeadm-bootstrap kubeadm-control-plane
```
[the remark this answers is not in the document]
Comment (Prajyot-Parab) [reply Makefile:142]: IMO we should stick to long-form as they seems more easier to understand by looking at help section. (Also above change will lead to few more follow up target name changes)

**Comment by fabriziopandini** [comment]
@Prajyot-Parab thanks! this is a great first contribution!
/lgtm

@vincepri @CecileRobertMichon @enxebre @sbueringer PTAL when you have some spare time

File: Makefile [hunk 13]
```diff
@@ -455,72 +422,63 @@ docker-build-kubeadm-control-plane: ## Build the docker image for kubeadm contro
...
-docker-push-all: $(addprefix docker-push-,$(ALL_ARCH))
-	$(MAKE) docker-push-core-manifest
-	$(MAKE) docker-push-kubeadm-bootstrap-manifest
-	$(MAKE) docker-push-kubeadm-control-plane-manifest
+KUBEBUILDER_ASSETS ?= $(shell $(SETUP_ENVTEST) use --use-env -p path $(KUBEBUILDER_ENVTEST_KUBERNETES_VERSION))
 
-docker-push-%:
-	$(MAKE) ARCH=$* docker-push
+.PHONY: test
+test: $(SETUP_ENVTEST) ## Run all the tests
```
Comment (CecileRobertMichon) [inline Makefile:440]:
```suggestion
test: $(SETUP_ENVTEST) ## Run unit and integration tests
```

The description "Run all the tests" may be misleading as it's not actually running ALL tests but runs all unit and integration tests in the repo, ie. it doesn't run e2e tests.

https://cluster-api.sigs.k8s.io/developer/testing.html#running-unit-and-integration-tests

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm aside from minor comment on the test target

File: Makefile (hunk 13 above)
Comment (Prajyot-Parab) [reply Makefile]: Done.

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
@fabriziopandini should we merge this or did you want others to look at it?

This is going to cause conflicts with any PR that updates the Makefile so the sooner it merges the better

**Comment by fabriziopandini** [comment]
> @fabriziopandini should we merge this or did you want others to look at it?
>
> This is going to cause conflicts with any PR that updates the Makefile so the sooner it merges the better

Let's get this moving, this PR greatly improves makefile usability

/lgtm
/approve

**Pull request merged**